include/swift/Basic/ExperimentalDependencies.h

```diff
@@ -0,0 +1,968 @@
+  DependencyKey(NodeKind kind, DeclAspect aspect, std::string context,
+                std::string name)
+      : kind(kind), aspect(aspect), context(context), name(name) {
+    assert(verify());
+  }
+
+  NodeKind getKind() const { return kind; }
+  DeclAspect getAspect() const { return aspect; }
+  StringRef getContext() const { return context; }
+  StringRef getName() const { return name; }
+
+  StringRef getSwiftDepsFromSourceFileProvide() const {
+    assert(getKind() == NodeKind::sourceFileProvide &&
+           "Receiver must be sourceFileProvide.");
+    return getName();
+  }
+
+  bool operator==(const DependencyKey &rhs) const {
+    return getKind() == rhs.getKind() && getAspect() == rhs.getAspect() &&
+           getContext() == rhs.getContext() && getName() == rhs.getName();
+  }
+
+  bool operator!=(const DependencyKey &rhs) const { return !(*this == rhs); }
+
+  size_t hash() const {
+    return std::hash<size_t>()(size_t(kind)) ^
+           std::hash<size_t>()(size_t(aspect) * size_t(NodeKind::kindCount)) ^
+           std::hash<std::string>()(name) ^ std::hash<std::string>()(context);
+  }
+  bool isImplementation() const {
+    return getAspect() == DeclAspect::implementation;
+  }
+  bool isInterface() const { return getAspect() == DeclAspect::interface; }
+
+  /// Given some type of provided entity compute the context field of the key.
+  template <NodeKind kind, typename Entity>
+  static std::string computeContextForProvdedEntity(Entity);
+
+  /// Given some type of provided entity compute the name field of the key.
+  template <NodeKind kind, typename Entity>
+  static std::string computeNameForProvidedEntity(Entity);
+
+  /// Given some type of depended-upon entity create the key.
+  template <NodeKind kind, typename Entity>
+  static DependencyKey createDependedUponKey(const Entity &);
+
+  std::string humanReadableName() const;
+
+  /// Convert to- or from- another format via the argument functions.
+  /// See \ref FrontendNode::serializeOrDeserialize.
+  void serializeOrDeserialize(function_ref<void(size_t &)> convertInt,
+                              function_ref<void(std::string &)> convertString) {
+    size_t k = size_t(kind);
+    convertInt(k);
+    kind = NodeKind(k);
+    size_t a = size_t(aspect);
+    convertInt(a);
+    aspect = DeclAspect(a);
+    convertString(name);
+    convertString(context);
+  }
+
+  void dump() const { llvm::errs() << std::string(*this) << "\n"; }
+
+  /// For debugging, see \ref TwoStageMap::verify
+  operator std::string() const;
+
+  bool verify() const;
+
+  /// Since I don't have Swift enums, ensure name corresspondence here.
+  static void verifyNodeKindNames();
+
+  /// Since I don't have Swift enums, ensure name corresspondence here.
+  static void verifyDeclAspectNames();
+
+private:
+  // Name conversion helpers
+
+  /// Represent an aggregate for some kinds as a mangled type:
+  static std::string mangleTypeAsContext(const NominalTypeDecl *);
```

Comment:
Layering violation: you can't depend on AST from Basic.

Reply:
Sigh... I want to keep the construction and deconstruction of DependencyKeys together, so I'll pass in a function from the front end.